File: exercism/ruby/pig-latin/pig_latin.rb
class PigLatin

  def self.translate(text)
    text.split.map{|word| translate_word(word) }.join(" ")
  end

private

  def self.translate_word(word)
    index = case
    when /^([bcdfghlklmnpqrstvwxyz]qu|thr|sch)/ =~ word then 3
    when /^(ch|qu|th)/ =~ word then 2
    when /^(ye|[bcdfghlklmnpqrstvwz]|[xy][aeiou])/ =~ word then 1
    else 0
    end
    word.chars.rotate(index).join + "ay"
  end
end

File: exercism/kotlin/pig-latin/src/main/kotlin/PigLatin.kt
object PigLatin {
  fun translate(text: String): String =
    text.split(' ').map { translateWord(it) }.joinToString(" ")

  private fun translateWord(word: String): String {
    val index = when {
      word.matches(Regex("([bcdfghlklmnpqrstvwxyz]qu|thr|sch).*")) -> 3
      word.matches(Regex("(ch|qu|th).*")) -> 2
      word.matches(Regex("(ye|[bcdfghlklmnpqrstvwz]|[xy][aeiou]).*")) -> 1
      else -> 0
    }
    return word.substring(index, word.length) + word.substring(0, index) + "ay"
  }
} 
